File: bindings/typescript/src/lifecycle.ts
import { SigningCoordinator, WasmSigningAdapter } from "./internal/signing.js";
import { registerStatusSnapshot } from "./internal/lifecycle-resources.js";
import { AuthsWorkflowError } from "./workflow/errors.js";
import type {
  ApprovalConfiguration,
  ApprovalPolicyReference,
  Signer,
  WorkflowWasmEngine,
} from "./workflow.js";
import { loadPackagedWorkflowEngine } from "./verifier/wasm.js";

const DIGEST_TOKEN = Symbol("auths-protocol-digest");
const SIGNED_STATUS_TOKEN = Symbol("auths-signed-status");
const SNAPSHOT_TOKEN = Symbol("auths-status-snapshot");
const LIFECYCLE_TOKEN = Symbol("auths-lifecycle");

export type LifecycleState = "active" | "revoked" | "superseded";

export class ProtocolDigest<Kind extends string> {
  declare private readonly __kind: Kind;

  private constructor(token: typeof DIGEST_TOKEN, bytes: Uint8Array) {
    if (token !== DIGEST_TOKEN) throw new TypeError("sealed Auths protocol digest");
    protocolDigests.set(this, bytes);
    Object.freeze(this);
  }

  static parse<Kind extends string>(kind: Kind, value: string): ProtocolDigest<Kind> {
    if (!/^[0-9a-f]{64}$/.test(value)) {
      throw new AuthsWorkflowError("invalid-authority", `${kind} must be 64 lowercase hex characters`);
    }
    return new ProtocolDigest(DIGEST_TOKEN, Uint8Array.from(
      { length: 32 },
      (_, index) => Number.parseInt(value.slice(index * 2, index * 2 + 2), 16),
    ));
  }

}

const protocolDigests = new WeakMap<ProtocolDigest<string>, Uint8Array>();

function protocolDigestBytes(value: ProtocolDigest<string>): Uint8Array {
  const bytes = protocolDigests.get(value);
  if (bytes === undefined) throw new AuthsWorkflowError("invalid-authority", "protocol digest is forged");
  return bytes.slice();
}

export type GrantId = ProtocolDigest<"grant-id">;
export type StatusSnapshotId = ProtocolDigest<"status-snapshot-id">;
export type EvidenceId = ProtocolDigest<"evidence-id">;

export const grantId = (value: string): GrantId => ProtocolDigest.parse("grant-id", value);
export const statusSnapshotId = (value: string): StatusSnapshotId =>
  ProtocolDigest.parse("status-snapshot-id", value);
export const evidenceId = (value: string): EvidenceId => ProtocolDigest.parse("evidence-id", value);

export interface CriticalExtensionInput {
  readonly id: string;
  readonly bytes: Uint8Array;
}

export interface PrincipalStatusRequest {
  readonly method: string;
  readonly principal: string;
  readonly purpose: string;
  readonly state: LifecycleState;
  readonly sequence: bigint;
  readonly observedAt: bigint;
  readonly validUntil: bigint;
  readonly issuer: string;
  readonly extensions?: readonly CriticalExtensionInput[];
}

export interface GrantStatusRequest {
  readonly method: string;
  readonly grantId: GrantId;
  readonly state: LifecycleState;
  readonly sequence: bigint;
  readonly observedAt: bigint;
  readonly validUntil: bigint;
  readonly issuer: string;
  readonly extensions?: readonly CriticalExtensionInput[];
}

export interface LifecycleSigningOptions {
  readonly signer: Signer;
  readonly approval: ApprovalConfiguration;
  readonly requiredApproval: ApprovalPolicyReference;
  readonly expiresAt?: bigint;
}

interface SignedStatusResources {
  readonly bytes: Uint8Array;
}

const signedStatusResources = new WeakMap<object, SignedStatusResources>();

export class SignedPrincipalStatus {
  declare private readonly __principalStatus: void;

  private constructor(token: typeof SIGNED_STATUS_TOKEN, bytes: Uint8Array) {
    if (token !== SIGNED_STATUS_TOKEN) throw new TypeError("sealed Auths principal status");
    signedStatusResources.set(this, { bytes });
    Object.freeze(this);
  }

  static create(token: typeof SIGNED_STATUS_TOKEN, bytes: Uint8Array): SignedPrincipalStatus {
    return new SignedPrincipalStatus(token, bytes);
  }
}

export class SignedGrantStatus {
  declare private readonly __grantStatus: void;

  private constructor(token: typeof SIGNED_STATUS_TOKEN, bytes: Uint8Array) {
    if (token !== SIGNED_STATUS_TOKEN) throw new TypeError("sealed Auths grant status");
    signedStatusResources.set(this, { bytes });
    Object.freeze(this);
  }

  static create(token: typeof SIGNED_STATUS_TOKEN, bytes: Uint8Array): SignedGrantStatus {
    return new SignedGrantStatus(token, bytes);
  }
}

export interface StatusTrustRule {
  readonly method: string;
  readonly issuer: string;
  readonly sequenceFloor: bigint;
}

export interface PrincipalStatusSnapshotRequest {
  readonly id: StatusSnapshotId;
  readonly observedAt: bigint;
  readonly validUntil: bigint;
  readonly statements: readonly SignedPrincipalStatus[];
  readonly checkpoints?: readonly EvidenceId[];
  readonly trust?: readonly StatusTrustRule[];
}

export interface GrantStatusSnapshotRequest {
  readonly id: StatusSnapshotId;
  readonly observedAt: bigint;
  readonly validUntil: bigint;
  readonly statements: readonly SignedGrantStatus[];
  readonly checkpoints?: readonly EvidenceId[];
  readonly trust?: readonly StatusTrustRule[];
}

export class PrincipalStatusSnapshot {
  declare private readonly __principalSnapshot: void;
  readonly id: StatusSnapshotId;
  readonly statementCount: number;

  private constructor(
    token: typeof SNAPSHOT_TOKEN,
    id: StatusSnapshotId,
    statementCount: number,
    cbor: Uint8Array,
  ) {
    if (token !== SNAPSHOT_TOKEN) throw new TypeError("sealed Auths principal status snapshot");
    this.id = id;
    this.statementCount = statementCount;
    registerStatusSnapshot(this, cbor);
    Object.freeze(this);
  }

  static create(
    token: typeof SNAPSHOT_TOKEN,
    id: StatusSnapshotId,
    statementCount: number,
    cbor: Uint8Array,
  ): PrincipalStatusSnapshot {
    return new PrincipalStatusSnapshot(token, id, statementCount, cbor);
  }
}

export class GrantStatusSnapshot {
  declare private readonly __grantSnapshot: void;
  readonly id: StatusSnapshotId;
  readonly statementCount: number;

  private constructor(
    token: typeof SNAPSHOT_TOKEN,
    id: StatusSnapshotId,
    statementCount: number,
    cbor: Uint8Array,
  ) {
    if (token !== SNAPSHOT_TOKEN) throw new TypeError("sealed Auths grant status snapshot");
    this.id = id;
    this.statementCount = statementCount;
    registerStatusSnapshot(this, cbor);
    Object.freeze(this);
  }

  static create(
    token: typeof SNAPSHOT_TOKEN,
    id: StatusSnapshotId,
    statementCount: number,
    cbor: Uint8Array,
  ): GrantStatusSnapshot {
    return new GrantStatusSnapshot(token, id, statementCount, cbor);
  }
}

export class LifecycleAuthor {
  readonly #engine: WorkflowWasmEngine;
  #active = true;

  private constructor(token: typeof LIFECYCLE_TOKEN, engine: WorkflowWasmEngine) {
    if (token !== LIFECYCLE_TOKEN) throw new TypeError("sealed Auths lifecycle author");
    this.#engine = engine;
  }

  static create(token: typeof LIFECYCLE_TOKEN, engine: WorkflowWasmEngine): LifecycleAuthor {
    return new LifecycleAuthor(token, engine);
  }

  async authorPrincipalStatus(
    request: PrincipalStatusRequest,
    options: LifecycleSigningOptions,
  ): Promise<SignedPrincipalStatus> {
    this.#assertActive();
    const unsigned = this.#engine.encodePrincipalStatusStatementV1(
      request.method,
      request.principal,
      request.purpose,
      request.state,
      request.sequence,
      request.observedAt,
      request.validUntil,
      request.issuer,
      copyExtensions(request.extensions),
    );
    const bytes = await this.#sign("principal-status", request.issuer, request.validUntil, unsigned, options);
    return SignedPrincipalStatus.create(SIGNED_STATUS_TOKEN, bytes);
  }

  async authorGrantStatus(
    request: GrantStatusRequest,
    options: LifecycleSigningOptions,
  ): Promise<SignedGrantStatus> {
    this.#assertActive();
    const unsigned = this.#engine.encodeGrantStatusStatementV1(
      request.method,
      protocolDigestBytes(request.grantId),
      request.state,
      request.sequence,
      request.observedAt,
      request.validUntil,
      request.issuer,
      copyExtensions(request.extensions),
    );
    const bytes = await this.#sign("grant-status", request.issuer, request.validUntil, unsigned, options);
    return SignedGrantStatus.create(SIGNED_STATUS_TOKEN, bytes);
  }

  principalSnapshot(request: PrincipalStatusSnapshotRequest): PrincipalStatusSnapshot {
    this.#assertActive();
    const native = this.#engine.parsePrincipalStatusSnapshotV1(snapshotInput(request));
    try {
      return PrincipalStatusSnapshot.create(
        SNAPSHOT_TOKEN,
        request.id,
        native.statementCount,
        new Uint8Array(native.cbor),
      );
    } finally {
      native.free?.();
    }
  }

  grantSnapshot(request: GrantStatusSnapshotRequest): GrantStatusSnapshot {
    this.#assertActive();
    const native = this.#engine.parseGrantStatusSnapshotV1(snapshotInput(request));
    try {
      return GrantStatusSnapshot.create(
        SNAPSHOT_TOKEN,
        request.id,
        native.statementCount,
        new Uint8Array(native.cbor),
      );
    } finally {
      native.free?.();
    }
  }

  dispose(): void {
    this.#active = false;
  }

  async #sign(
    objectKind: "principal-status" | "grant-status",
    issuer: string,
    validUntil: bigint,
    unsignedObject: Uint8Array,
    options: LifecycleSigningOptions,
  ): Promise<Uint8Array> {
    const principal = await options.signer.publicIdentity();
    if (principal.principal !== issuer) {
      throw new AuthsWorkflowError("signer-response-mismatch", "status issuer does not match signer principal");
    }
    const coordinator = new SigningCoordinator(new WasmSigningAdapter(this.#engine));
    const signed = await coordinator.execute({
      objectKind,
      unsignedObject,
      principal,
      signer: options.signer,
      approval: options.approval,
      requiredApproval: options.requiredApproval,
      expiresAt: options.expiresAt ?? validUntil,
      display: Object.freeze([
        Object.freeze({ label: "Object", value: objectKind }),
        Object.freeze({ label: "Issuer", value: issuer }),
      ]),
    });
    return signed.signedObject.slice();
  }

  #assertActive(): void {
    if (!this.#active) throw new AuthsWorkflowError("disposed", "lifecycle author is disposed");
  }
}

export async function loadLifecycleAuthor(): Promise<LifecycleAuthor> {
  return LifecycleAuthor.create(LIFECYCLE_TOKEN, await loadPackagedWorkflowEngine());
}

function copyExtensions(value: readonly CriticalExtensionInput[] | undefined): readonly object[] {
  return Object.freeze((value ?? []).map((extension) => Object.freeze({
    id: extension.id,
    bytes: extension.bytes.slice(),
  })));
}

function snapshotInput(
  request: PrincipalStatusSnapshotRequest | GrantStatusSnapshotRequest,
): object {
  return Object.freeze({
    id: protocolDigestBytes(request.id),
    observedAt: request.observedAt,
    validUntil: request.validUntil,
    statements: Object.freeze(request.statements.map((statement) => {
      const resources = signedStatusResources.get(statement);
      if (resources === undefined) {
        throw new AuthsWorkflowError("invalid-authority", "status statement is forged");
      }
      return resources.bytes.slice();
    })),
    checkpoints: Object.freeze((request.checkpoints ?? []).map(protocolDigestBytes)),
    trust: Object.freeze((request.trust ?? []).map((rule) => Object.freeze({ ...rule }))),
  });
}

File: bindings/python/python/auths/lifecycle.py
"""Typed principal and delegated-authority lifecycle authoring."""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Literal, Optional, Protocol, Sequence, Tuple, Union, runtime_checkable

from . import _native as native
from ._native import SignedObject
from .workflow import (
    ApprovalConfiguration,
    ApprovalPolicyReference,
    AuthsWorkflowError,
    Principal,
    ReviewField,
    Signer,
    _SigningCoordinator,
    _call_public_identity,
)

LifecycleState = Literal["active", "revoked", "superseded"]
_SIGNED_TOKEN = object()
_SNAPSHOT_TOKEN = object()


@dataclass(frozen=True)
class ProtocolDigest:
    bytes: bytes

    def __post_init__(self) -> None:
        value = bytes(self.bytes)
        if len(value) != 32:
            raise ValueError("protocol digest must contain 32 bytes")
        object.__setattr__(self, "bytes", value)

    @classmethod
    def parse(cls, value: str) -> ProtocolDigest:
        if len(value) != 64 or any(character not in "0123456789abcdef" for character in value):
            raise ValueError("protocol digest must be 64 lowercase hexadecimal characters")
        return cls(bytes.fromhex(value))


@dataclass(frozen=True)
class CriticalExtension:
    id: str
    bytes: bytes

    def __post_init__(self) -> None:
        object.__setattr__(self, "bytes", bytes(self.bytes))


@dataclass(frozen=True)
class PrincipalStatusRequest:
    method: str
    principal: Principal
    purpose: str
    state: LifecycleState
    sequence: int
    observed_at: int
    valid_until: int
    issuer: Principal
    extensions: Tuple[CriticalExtension, ...] = ()

    def __post_init__(self) -> None:
        object.__setattr__(self, "extensions", tuple(self.extensions))


@dataclass(frozen=True)
class GrantStatusRequest:
    method: str
    grant_id: ProtocolDigest
    state: LifecycleState
    sequence: int
    observed_at: int
    valid_until: int
    issuer: Principal
    extensions: Tuple[CriticalExtension, ...] = ()

    def __post_init__(self) -> None:
        object.__setattr__(self, "extensions", tuple(self.extensions))


@dataclass(frozen=True)
class IdentityRotation:
    previous: PrincipalStatusRequest
    current: PrincipalStatusRequest


class SignedPrincipalStatus:
    def __init__(self, token: object, value: SignedObject) -> None:
        if token is not _SIGNED_TOKEN:
            raise TypeError("sealed Auths principal status")
        self._value = value


class SignedGrantStatus:
    def __init__(self, token: object, value: SignedObject) -> None:
        if token is not _SIGNED_TOKEN:
            raise TypeError("sealed Auths grant status")
        self._value = value


@dataclass(frozen=True)
class StatusTrustRule:
    method: str
    issuer: Principal
    sequence_floor: int


class PrincipalStatusSnapshot:
    def __init__(
        self, token: object, identifier: ProtocolDigest, native: native.StatusSnapshot
    ) -> None:
        if token is not _SNAPSHOT_TOKEN:
            raise TypeError("sealed Auths principal status snapshot")
        self.id = identifier
        self._native = native


class GrantStatusSnapshot:
    def __init__(
        self, token: object, identifier: ProtocolDigest, native: native.StatusSnapshot
    ) -> None:
        if token is not _SNAPSHOT_TOKEN:
            raise TypeError("sealed Auths grant status snapshot")
        self.id = identifier
        self._native = native


class LifecycleAuthor:
    def __init__(
        self,
        *,
        signer: Signer,
        approval: ApprovalConfiguration,
        required_approval: ApprovalPolicyReference,
    ) -> None:
        self._signer = signer
        self._approval = approval
        self._required_approval = required_approval
        self._closed = False

    async def principal_status(
        self, request: PrincipalStatusRequest
    ) -> SignedPrincipalStatus:
        self._assert_open()
        if type(request) is not PrincipalStatusRequest:
            raise TypeError("request must be a PrincipalStatusRequest")
        unsigned = native.principal_status_statement(
            request.method,
            request.principal,
            request.purpose,
            request.state,
            request.sequence,
            request.observed_at,
            request.valid_until,
            request.issuer,
            [(value.id, value.bytes) for value in request.extensions],
        )
        return SignedPrincipalStatus(
            _SIGNED_TOKEN,
            await self._sign(unsigned, request.issuer, request.valid_until, "Principal status"),
        )

    async def grant_status(self, request: GrantStatusRequest) -> SignedGrantStatus:
        self._assert_open()
        if type(request) is not GrantStatusRequest:
            raise TypeError("request must be a GrantStatusRequest")
        unsigned = native.grant_status_statement(
            request.method,
            request.grant_id.bytes,
            request.state,
            request.sequence,
            request.observed_at,
            request.valid_until,
            request.issuer,
            [(value.id, value.bytes) for value in request.extensions],
        )
        return SignedGrantStatus(
            _SIGNED_TOKEN,
            await self._sign(unsigned, request.issuer, request.valid_until, "Grant status"),
        )

    def close(self) -> None:
        self._closed = True

    async def _sign(
        self, unsigned: native.UnsignedObject, issuer: Principal, expires_at: int, label: str
    ) -> SignedObject:
        descriptor = await _call_public_identity(self._signer, "lifecycle signer")
        if descriptor.principal.value != issuer.value:
            raise AuthsWorkflowError(
                "invalid-principal", "lifecycle signer does not control the declared issuer"
            )
        result = await _SigningCoordinator().execute(
            unsigned=unsigned,
            principal=descriptor,
            signer=self._signer,
            approval=self._approval,
            required_approval=self._required_approval,
            expires_at=expires_at,
            display=(ReviewField("Operation", label), ReviewField("Issuer", issuer.value)),
        )
        return result.signed_object

    def _assert_open(self) -> None:
        if self._closed:
            raise AuthsWorkflowError("disposed", "lifecycle author is closed")


def principal_status_snapshot(
    identifier: ProtocolDigest,
    *,
    observed_at: int,
    valid_until: int,
    statements: Sequence[SignedPrincipalStatus],
    checkpoints: Sequence[ProtocolDigest] = (),
    trust: Sequence[StatusTrustRule] = (),
) -> PrincipalStatusSnapshot:
    values = tuple(statements)
    if any(type(value) is not SignedPrincipalStatus for value in values):
        raise TypeError("principal snapshot contains another status kind")
    snapshot = native.status_snapshot(
        "principal",
        identifier.bytes,
        observed_at,
        valid_until,
        [value._value for value in values],
        [value.bytes for value in checkpoints],
        [(value.method, value.issuer.value, value.sequence_floor) for value in trust],
    )
    return PrincipalStatusSnapshot(_SNAPSHOT_TOKEN, identifier, snapshot)


def grant_status_snapshot(
    identifier: ProtocolDigest,
    *,
    observed_at: int,
    valid_until: int,
    statements: Sequence[SignedGrantStatus],
    checkpoints: Sequence[ProtocolDigest] = (),
    trust: Sequence[StatusTrustRule] = (),
) -> GrantStatusSnapshot:
    values = tuple(statements)
    if any(type(value) is not SignedGrantStatus for value in values):
        raise TypeError("grant snapshot contains another status kind")
    snapshot = native.status_snapshot(
        "grant",
        identifier.bytes,
        observed_at,
        valid_until,
        [value._value for value in values],
        [value.bytes for value in checkpoints],
        [(value.method, value.issuer.value, value.sequence_floor) for value in trust],
    )
    return GrantStatusSnapshot(_SNAPSHOT_TOKEN, identifier, snapshot)


def withdraw_delegation(
    *,
    method: str,
    grant_id: ProtocolDigest,
    issuer: Principal,
    sequence: int,
    valid_for: int,
    observed_at: Optional[int] = None,
) -> GrantStatusRequest:
    observed = int(time.time()) if observed_at is None else observed_at
    return GrantStatusRequest(
        method, grant_id, "revoked", sequence, observed, observed + valid_for, issuer
    )


def record_compromise(
    *,
    method: str,
    principal: Principal,
    purpose: str,
    issuer: Principal,
    sequence: int,
    valid_for: int,
    observed_at: Optional[int] = None,
) -> PrincipalStatusRequest:
    observed = int(time.time()) if observed_at is None else observed_at
    return PrincipalStatusRequest(
        method,
        principal,
        purpose,
        "revoked",
        sequence,
        observed,
        observed + valid_for,
        issuer,
    )


def rotate_identity(
    *,
    method: str,
    previous: Principal,
    current: Principal,
    purpose: str,
    issuer: Principal,
    previous_sequence: int,
    current_sequence: int,
    valid_for: int,
    observed_at: Optional[int] = None,
) -> IdentityRotation:
    observed = int(time.time()) if observed_at is None else observed_at
    valid_until = observed + valid_for
    return IdentityRotation(
        PrincipalStatusRequest(
            method,
            previous,
            purpose,
            "superseded",
            previous_sequence,
            observed,
            valid_until,
            issuer,
        ),
        PrincipalStatusRequest(
            method,
            current,
            purpose,
            "active",
            current_sequence,
            observed,
            valid_until,
            issuer,
        ),
    )


StatusSnapshot = Union[PrincipalStatusSnapshot, GrantStatusSnapshot]


@runtime_checkable
class StatusProvider(Protocol):
    contract_version: int

    async def principal(
        self, identifier: ProtocolDigest, *, observed_at: int
    ) -> PrincipalStatusSnapshot: ...

    async def grant(
        self, identifier: ProtocolDigest, *, observed_at: int
    ) -> GrantStatusSnapshot: ...

__all__ = [
    "CriticalExtension",
    "GrantStatusRequest",
    "GrantStatusSnapshot",
    "IdentityRotation",
    "LifecycleAuthor",
    "LifecycleState",
    "PrincipalStatusRequest",
    "PrincipalStatusSnapshot",
    "ProtocolDigest",
    "SignedGrantStatus",
    "SignedPrincipalStatus",
    "StatusSnapshot",
    "StatusProvider",
    "StatusTrustRule",
    "grant_status_snapshot",
    "principal_status_snapshot",
    "record_compromise",
    "rotate_identity",
    "withdraw_delegation",
]
